### PR TITLE
Fix SO3/SE3Tangent random

### LINKED ISSUE #180: Broken exp().log() in SO3/SE3
The following test is broken in `SO3/SE3`:
```cpp
delta == delta.exp().log();
```
It fails roughly as often is passes. The test that was supposed to check this was not run (my bad for not fixing a copy/paste).

Note that `SE3` relies on `SO3` for a part, so most likely the issue comes from `SO3`.
Note also that `state.log().exp()` seems to work fine and is properly tested in CI.

I added a minimal reproducible example [here on branch `fix/exp_log`](https://github.com/artivis/manif/commit/6b499719f910ee17e739b79b6914acf5501264e8#diff-80b600a72e65d107d204a378d72aeaa0206a3f47d24fa3e810d0efa1dd8de590).

### PR DESCRIPTION
**Pull request by artivis**
Implement the random point in sphere as per [this SO answer](https://stackoverflow.com/a/5408843/9709397).

Closes #180.

### CODE REVIEW

File: include/manif/impl/eigen.h [hunk 1]
```diff
@@ -163,6 +163,33 @@ skew(const Eigen::MatrixBase<_Derived>& v)
             -v(1),   +v(0),    T(0.) ).finished();
 }
 
+template <typename Scalar>
+Eigen::Matrix<Scalar, 3, 1> generate_point_in_sphere(Scalar radius)
```
[the remark this answers is not in the document]
Comment (joansola) [reply include/manif/impl/eigen.h:167]: Why the underscores and not camelCase?

**Review by joansola — COMMENTED** [review]
A number of things, from naming to code optimizations.

File: include/manif/impl/eigen.h [hunk 2]
```diff
@@ -163,6 +163,33 @@ skew(const Eigen::MatrixBase<_Derived>& v)
...
+  // random(-1, 1)
+  Scalar costheta = Scalar(-1) + static_cast<Scalar>(rand()) / (static_cast<Scalar>(RAND_MAX/(Scalar(1) - Scalar(-1))));
+  // random(0, 1)
+  Scalar u = static_cast<Scalar>(rand()) / static_cast<Scalar>(RAND_MAX);
+
+  Scalar theta = acos(costheta);
+  Scalar r = radius * cbrt(u);
+
+  return Eigen::Matrix<Scalar, 3, 1>(
+    r * sin(theta) * cos(phi),
```
[the remark this answers is not in the document]
Comment (joansola) [reply include/manif/impl/eigen.h:187]: You can define a `tmp = r*sin(theta)` and save a sin() and a product.

[the remark this answers is not in the document]
Comment (joansola) [reply include/manif/impl/eigen.h:187]: some systems even admit the sincos() function that saves time whenever you need both sin and cos of the same angle.

File: include/manif/impl/eigen.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (artivis) [reply include/manif/impl/eigen.h:167]: > also, I think this function name deserves the keyword "random" somewhere, no? e.g. `random_point_in_ball(radius)`

How about `randPointInBall`?

> Why the underscores and not camelCase?

It is true that I have mixed a bit both in the library. I think all member functions are following the camelCase standard tho. Another ticket to open ;D

[the remark this answers is not in the document]
Comment (joansola) [reply include/manif/impl/eigen.h:167]: OK for `randPointInBall`

File: include/manif/impl/eigen.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (artivis) [reply include/manif/impl/eigen.h:187]: Will do. 
`sincos` is not standard C nor C++ unfortunately.

File: include/manif/impl/se_2_3/SE_2_3Tangent_base.h [hunk 3]
```diff
@@ -390,8 +390,8 @@ struct RandomEvaluatorImpl<SE_2_3TangentBase<Derived>>
 {
   static void run(SE_2_3TangentBase<Derived>& m)
   {
-    m.coeffs().setRandom();                // in [-1,1]
-    m.coeffs().template segment<3>(3) *= MANIF_PI; // in [-PI,PI]
+    m.coeffs().setRandom(); // in [-1,1]
+    m.coeffs().template segment<3>(3) = randPointInBall(MANIF_PI); // in [-PI,PI]
```
Comment (joansola) [inline include/manif/impl/se_2_3/SE_2_3Tangent_base.h:394]: Fix comment to `// In ball of radius PI`

**Review by joansola — CHANGES_REQUESTED** [review]
almost there

File: include/manif/impl/se_2_3/SE_2_3Tangent_base.h (hunk 3 above)
Comment (artivis) [reply include/manif/impl/se_2_3/SE_2_3Tangent_base.h]: Done :+1:

**Review by joansola — APPROVED** [review]
OK!

**Pull request merged**